src/go/pt-galera-log-explainer/internal.go

```diff
@@ -0,0 +1,218 @@
+		grep is everywhere, grep is good enough, it even enable to use the stdout pipe.
+
+		The usual bottleneck with grep is that it is single-threaded, but we actually benefit
+		from a sequential scan here as we will rely on the log order.
+
+		Also, being sequential also ensure this program is light enough to run without too much impacts
+		It also helps to be transparent and not provide an obscure tool that work as a blackbox
+	*/
+	if runtime.GOOS == "darwin" && CLI.GrepCmd == "grep" {
+		logger.Warn().Msg("On Darwin systems, use 'pt-galera-log-explainer --grep-cmd=ggrep' as it requires grep v3")
+	}
+
+	cmd := exec.Command(CLI.GrepCmd, CLI.GrepArgs, compiledRegex, path)
+
+	out, _ := cmd.StdoutPipe()
+	defer out.Close()
+
+	err := cmd.Start()
+	if err != nil {
+		return errors.Wrapf(err, "failed to search in %s", path)
+	}
+
+	// grep treatment
+	s := bufio.NewScanner(out)
+	for s.Scan() {
+		stdout <- s.Text()
+	}
+
+	// double-check it stopped correctly
+	if err = cmd.Wait(); err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok && exiterr.ExitCode() == 1 {
+			return nil
+		}
+		return errors.Wrap(err, "grep subprocess error")
+	}
+
+	return nil
+}
+
+func sanitizeLine(s string) string {
+	if len(s) > 0 && s[0] == '\t' {
+		return s[1:]
+	}
+	return s
+}
+
+// iterateOnGrepResults will take line by line each logs that matched regex
+// it will iterate on every regexes in slice, and apply the handler for each
+// it also filters out --since and --until rows
+func iterateOnGrepResults(path string, regexes types.RegexMap, grepStdout <-chan string) types.LocalTimeline {
+
+	var (
+		lt        types.LocalTimeline
+		displayer types.LogDisplayer
+		timestamp time.Time
+	)
+	ctx := types.NewLogCtx()
+	ctx.FilePath = path
+
+	for line := range grepStdout {
+		line = sanitizeLine(line)
+
+		var date *types.Date
+		t, layout, ok := regex.SearchDateFromLog(line)
+		if ok {
+			// diff between date and timestamp:
+			// timestamp is an internal usage to handle translations, it must be non-empty
+			// date is something that will be displayed ultimately, it can empty
+			date = types.NewDate(t, layout)
+			timestamp = t
+		} // else, keep the previous timestamp
+
+		// If it's recentEnough, it means we already validated a log: every next logs necessarily happened later
+		// this is useful because not every logs have a date attached, and some without date are very useful
+		if CLI.Since != nil && CLI.Since.After(timestamp) {
+			continue
+		}
+		if CLI.Until != nil && CLI.Until.Before(timestamp) {
+			return lt
+		}
+
+		filetype := regex.FileType(line, CLI.PxcOperator)
+		ctx.FileType = filetype
+
+		// We have to find again what regex worked to get this log line
+		// it can match multiple regexes
+		for key, regex := range regexes {
+			if !regex.Regex.MatchString(line) || utils.SliceContains(CLI.ExcludeRegexes, key) {
```

Comment:
fyi, it's by far the current bottleneck (around 35%-40% of total CPU time spent on medium files in regression tests with pprof)
My idea: make a quick estimation using "strings.Contains(line, word)" for every word in the regexes. So that it's strict comparison, and not regex anymore which is too slow
Few false positives would be okay anyway, as regexes are re-checked with "internalRegex" in the next line in "regex.Handle"